### PR TITLE
[SPARK-22037][SQL] Collapse Project if it is the child of Aggregate

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

If Aggregate's child is Project, collapse the Project into the Aggregate.
This rule is to make query plans simpler.

## How was this patch tested?

Unit test

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -541,6 +541,25 @@ object CollapseProject extends Rule[LogicalPlan] {
         agg.copy(aggregateExpressions = buildCleanedProjectList(
           p.projectList, agg.aggregateExpressions))
       }
+    case agg @ Aggregate(groupingExpressions, aggregateExpressions, p: Project) =>
+      if (haveCommonNonDeterministicOutput(groupingExpressions, p.projectList) ||
+        haveCommonNonDeterministicOutput(aggregateExpressions, p.projectList)) {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This condition is not right.

**Comment by gatorsmile** [comment]
In whole-stage codegen, this will not gain any perf gain if we can remove the useless Project.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This condition is not right

**Comment by viirya** [comment]
The removed `Project` might affect the rule `PhysicalOperation` and reduce the chance to prune columns when reading relations.

**Comment by gengliangwang** [comment]
@gatorsmile @viirya  Thanks. I should close this PR.

**Pull request closed without merging**